node/network/approval-distribution/src/lib.rs

```diff
@@ -0,0 +1,839 @@
+				return;
+			}
+
+			// check if our knowledge of the peer already contains this assignment
+			match entry.known_by.entry(peer_id.clone()) {
+				hash_map::Entry::Occupied(knowledge) => {
+					if knowledge.get().known_messages.contains(&fingerprint) {
+						modify_reputation(ctx, peer_id, COST_DUPLICATE_MESSAGE).await;
+						return;
+					}
+				}
+				hash_map::Entry::Vacant(_) => {
+					modify_reputation(ctx, peer_id.clone(), COST_UNEXPECTED_MESSAGE).await;
+				}
+			}
+
+			// if the assignment is known to be valid, reward the peer
+			if entry.knowledge.known_messages.contains(&fingerprint) {
+				modify_reputation(ctx, peer_id.clone(), BENEFIT_VALID_MESSAGE).await;
+				entry.known_by.entry(peer_id).or_default().known_messages.insert(fingerprint.clone());
+				return;
+			}
+
+			// FIXME: possibly deadlocks due to https://github.com/paritytech/polkadot/issues/2149
+			let (tx, rx) = oneshot::channel();
+
+			ctx.send_message(AllMessages::ApprovalVoting(ApprovalVotingMessage::CheckAndImportApproval(
+				vote.clone(),
+				tx,
+			))).await;
+
+			let result = match rx.await {
+				Ok(result) => result,
+				Err(_) => {
+					tracing::debug!(
+						target: LOG_TARGET,
+						"The approval voting subsystem is down",
+					);
+					return;
+				}
+			};
+
+			match result {
+				ApprovalCheckResult::Accepted => {
+					modify_reputation(ctx, peer_id.clone(), BENEFIT_VALID_MESSAGE_FIRST).await;
+
+					entry.knowledge.known_messages.insert(fingerprint.clone());
+					entry.known_by
+						.entry(peer_id)
+						.or_default()
+						.known_messages
+						.insert(fingerprint.clone());
+				}
+				ApprovalCheckResult::Bad => {
+					modify_reputation(ctx, peer_id, COST_INVALID_MESSAGE).await;
+					return;
+				}
+			}
+		} else {
+			entry.knowledge.known_messages.insert(fingerprint.clone());
+		}
+
+		match entry.candidates.get_mut(&candidate_index) {
+			Some(candidate_entry) => {
+				// set the approval state for validator_index to Approved
+				// it should be in assigned state already
+				match candidate_entry.approvals.remove(&validator_index) {
+					Some(ApprovalState::Assigned(cert)) => {
+						candidate_entry.approvals.insert(
+							validator_index,
+							ApprovalState::Approved(cert, vote.signature.clone()),
+						);
+					}
+					_ => {
+						tracing::warn!(
+							target: LOG_TARGET,
+							hash = ?block_hash,
+							?candidate_index,
+							"Expected a candidate entry with `ApprovalState::Assigned`",
+						);
+					}
+				}
+			}
+			None => {
+				tracing::warn!(
+					target: LOG_TARGET,
+					hash = ?block_hash,
+					?candidate_index,
+					"Expected a candidate entry on import_and_circulate_approval",
+				);
+			}
+		}
+
+		// Dispatch a ApprovalDistributionV1Message::Approval(vote)
+		// to all peers in the BlockEntry's known_by set,
+		// excluding the peer in the source, if source has kind MessageSource::Peer.
+		let maybe_peer_id = source.peer_id();
+		let peers = self.peer_views
```

Comment:
likewise here, should exclude peers who already know the message

Reply:
it'd be good to test this (receiving a message for a second time sends only to peers who weren't aware).